### PR TITLE
NOT TO BE MERGED! Exploratory unit test for the unique filter to be reviewed

### PR DESCRIPTION
**Pull request by pkozlowski-opensource**
hi! Spent some time checking what is wrong with the the unique filter and ended up adding some test. Could you guys have a look at let me know if the assumptions in the tests are correct? As looking at the implementation I don't see how it could possibly work... I mean, given assumptions in my tests the implementation is clearly wrong... I guess I must be missing something obvious here...

Sending this PR to discuss tests / impl, it should NOT TO BE MERGED! as it will break the build (there are failing tests).

### CODE REVIEW

**Comment by ajoslin** [comment]
Looks good, Pawel.  You should also add a test for arrays of primitives:

``` javascript
it('should return unique primitives in arrays', function() {
  expect(uniqueFilter([1,2,1,3])).toEqual([1,2,3])
});
```

**Comment by pkozlowski-opensource** [comment]
@ajoslin, Andy, thnx for having a look. I agree that it should work for arrays of primitives, adding tests right now. Actually I'm a bit puzzled about the way the filtering key is passed to the filter - from what I could understand from the doc we can have 3 cases:
- [key] {string} the name of the attribute of each object to compare for uniqueness - this makes sense for arrays of objects only
- if the key is empty, the entire object will be compared - will work for both objects and primitives
- if the key === false then no filtering will be performed

Actually the option with the 'empty' key is a bit puzzling - what should be considered as 'empty'? An empty string? or undefined as in your example?

**Comment by ajoslin** [comment]
I guess empty means .length==0 or undefined, yeah.

**Comment by ajoslin** [comment]
And also, it should probably treat a string with no key parameter like an array:

``` javascript
it('should return unique characters in a string, if empty key parameter', function() {
  expect(uniqueFilter('abac')).toEqual('abc')
});
```

**Comment by ProLoser** [comment]
Key was the parameter passed to the filter. If it is undefined, meaning no parameter is passed, it compares entire objects. I realize the undefined vs false case might be weird.

**Comment by ProLoser** [comment]
The string as an array seems like a strange edge case to me.

File: modules/filters/unique/test/uniqueSpec.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
+describe('unique', function() {
+  var uniqueFilter;
+
+  beforeEach(module('ui.filters'));
+  beforeEach(inject(function($filter) {
+    uniqueFilter = $filter('unique');
+  }));
+
+  it('should return unique entries based on object equality', function () {
+    var arrayToFilter = [{key: 'value'}, {key: 'value2'}, {key: 'value'}];
```
Comment (petebacondarwin) [inline modules/filters/unique/test/uniqueSpec.js:10]: What exactly do you mean by "object equality"?  Obviously, here arrayToFilter[0] !== arrayToFilter[2] but some kind of deep equality check might return true.

File: modules/filters/unique/test/uniqueSpec.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  }));
+
+  it('should return unique entries based on object equality', function () {
+    var arrayToFilter = [{key: 'value'}, {key: 'value2'}, {key: 'value'}];
+    expect(uniqueFilter(arrayToFilter, '')).toEqual([{key: 'value'}, {key: 'value2'}]);
+  });
+
+  it('should return unique entries based on the key provided', function () {
+    var arrayToFilter = [{key: 'value', other: 'other1'}, {key: 'value2', other: 'other2'}, {key: 'value', other: 'other3'}];
+    expect(uniqueFilter(arrayToFilter, 'key')).toEqual([{key: 'value'}, {key: 'value2'}]);
```
Comment (petebacondarwin) [inline modules/filters/unique/test/uniqueSpec.js:21]: I would expect this case to return the whole unique objects from the array not just the matching key and value.

```
expect(uniqueFilter(arrayToFilter, 'key')).toEqual([{key: 'value', other: 'other1'}, {key: 'value2', other: 'other2'}]);
```

File: modules/filters/unique/test/uniqueSpec.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  });
+
+  it('should return unique entries based on the key provided for complex objects', function () {
+    var arrayToFilter = [{key: 'value'}, {key: 'value2'}, {key: 'value'}];
+    expect(uniqueFilter(arrayToFilter, 'key')).toEqual([{key: 'value'}, {key: 'value2'}]);
```
Comment (petebacondarwin) [inline modules/filters/unique/test/uniqueSpec.js]: This is not as complex as the previous expectation...

File: modules/filters/unique/test/uniqueSpec.js (hunk 1 above)
Comment (pkozlowski-opensource) [inline modules/filters/unique/test/uniqueSpec.js:10]: Yes, this is _the_ tricky part.  Depending on the behavior we want to have we can consider 2 objects o1 and o2 being equal if:
- `o1===o2`  - this is reference equality
- `angular.equals(o1, o2)` - this is "deep" equality
- other methods (`angular.toJson(o1)===angular.toJson(o2)` for example - but this is not reliable at all)

I've started implementing equality based on `angular.equals` - this will work fine for all object types but want be super-super fast - O(n*n) - unless anyone knows a good hash-generation function for objects.

File: modules/filters/unique/test/uniqueSpec.js (hunk 2 above)
Comment (pkozlowski-opensource) [inline modules/filters/unique/test/uniqueSpec.js:21]: yes, you are completely right, since I had many failing tests I didn't catch this obvious thing... Of course the test is wrong here.

File: modules/filters/unique/test/uniqueSpec.js (hunk 3 above)
Comment (pkozlowski-opensource) [inline modules/filters/unique/test/uniqueSpec.js]: yeh, good catch once again, I've changed description of the wrong test.... Corrected on my side already but didn't re-submit since started to look into this equality thing (see my previous comments).

File: modules/filters/unique/test/uniqueSpec.js (hunk 1 above)
Comment (ProLoser) [inline modules/filters/unique/test/uniqueSpec.js:10]: I had no idea about angular.equals()

**Comment by ajoslin** [comment]
(string as array case): Yes, but what else would you as a user expect it to do for `{{'ababc' | unique}}`?  Tests need to cover edge cases too

**Comment by ProLoser** [comment]
I would expect it to do nothing (or fail).

Under that same logic, what do you do if someone does `{{'abadc' | unique:'id'}}`

**Comment by ajoslin** [comment]
There are no repeated instances of 'id', so nothing would be taken out.

**Comment by pkozlowski-opensource** [comment]
OK, pushed a new commit that has both working unit tests and fixes for the filter itself. It will work correctly for arrays but might be not super-efficient for arrays without any duplication.

**Comment by pkozlowski-opensource** [comment]
OK, guys, thank you for all the comments, I've integrated most of them in the b406c1a. The thing I've left for later is filtering on strings - it is not super-difficult to do but since this logic might be confusing let's add support for strings in another commit if needed.

Since I've opened this PR mostly to get you comments I will close this one and will open a fresh one with one commit only so it can be easily merged.
